### PR TITLE
ImportC add `#define ident integer` manifest constants

### PR DESCRIPTION
**Pull request by WalterBright**
This adds initial support for recognizing `#define` patterns and making them available to D.

This one only recognizes the pattern `#define identifier integer` and adds an `enum ident = integer;` to the symbols accessible by importing the file.

It also only works if the `-dD` switch is passed to `cpp`, although the test case is a `.i` file so it is not necessary to capture the `#define`s. Capturing is only done for `cpp` anyway.

`#undef` is not yet supported.

This works by having the lexer recognize lines starting with `#define` or `#undef` and append them to `defines[]`. At the end of the C parse, the lexer is reconfigured to parse those lines, recognize patterns in them, and append any new symbols to `symbols[]`.

### CODE REVIEW

File: src/dmd/cparse.d [hunk 1]
```diff
@@ -5078,4 +5100,63 @@ final class CParser(AST) : Parser!AST
...
+            if (p[0 .. 7] == "#define")
+            {
+                p += 7;
+                scan(&n);
+                //printf("%s\n", n.toChars());
+                if (n.value == TOK.identifier)
+                {
+                    auto id = n.ident;
+                    scan(&n);
+                    if (n.value == TOK.int32Literal)
```
Comment (ibuclaw) [inline src/dmd/cparse.d:5135]: This entire PR will just end up being reverted when we start doing the more complex cases, might as well do it right from the start.

File: src/dmd/dmodule.d [hunk 2]
```diff
@@ -1001,7 +1001,8 @@ extern (C++) final class Module : Package
         {
             filetype = FileType.c;
 
-            scope p = new CParser!AST(this, buf, cast(bool) docfile, target.c);
+            OutBuffer defines;
```
Comment (ibuclaw) [inline src/dmd/dmodule.d:1004]: All the work is done in cparse, so I don't see the point of this.

File: src/dmd/lexer.d [hunk 3]
```diff
@@ -146,6 +146,31 @@ class Lexer
         }
     }
 
+    /**************************************
+     * Reset lexer to lex #define's
+     */
+    final void resetDefineLines(const(char)[] slice)
```
Comment (ibuclaw) [inline src/dmd/lexer.d:152]: Move to cparse, its only user.

File: src/dmd/lexer.d [hunk 4]
```diff
@@ -146,6 +146,31 @@ class Lexer
...
+        inTokenStringConstant = 0;
+        lastDocLine = 0;
+        scanloc = Loc("#defines", 1, 1);
+    }
+
+    /**********************************
+     * Set up for next #define line.
+     * p should be at start of next line.
+     */
+    final void nextDefineLine()
```
Comment (ibuclaw) [inline src/dmd/lexer.d:169]: Ditto.

File: src/dmd/lexer.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/lexer.d:152]: It accesses private Lexer data.

File: src/dmd/lexer.d (hunk 4 above)
Comment (WalterBright) [reply src/dmd/lexer.d:169]: ditto :-)

File: src/dmd/dmodule.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/dmodule.d:1004]: cl and sppn do not interleave the #define's into the .i file. It's dumped separately. I did it this way to accommodate both methods.

File: src/dmd/cparse.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/cparse.d:5135]: It'll still need the rest of the PR. The #defines still have to be captured.

**Comment by WalterBright** [comment]
@ibuclaw I don't mind if later you dump this for a better method. What I do want, however, is progress. This is progress (from a user point of view).

File: src/dmd/dmodule.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/dmodule.d:1004]: Future PRs will pass in a `defines` that has been initialized with the captured `#defines`.

**Comment by WalterBright** [comment]
My initial goal is to do the manifest constants. This will capture perhaps 90% of the `#define` usage.

File: src/dmd/lexer.d (hunk 3 above)
Comment (ibuclaw) [reply src/dmd/lexer.d:152]: Then make it protected? :-)

Comment (WalterBright) [reply src/dmd/lexer.d:157]: I never much cared for protected. Having a function to do it is better encapsulation.

File: src/dmd/cparse.d [hunk 5]
```diff
@@ -59,8 +59,16 @@ final class CParser(AST) : Parser!AST
      */
     Array!(void*) typedefTab;  /// Array of AST.Type[Identifier], typedef's indexed by Identifier
 
+    /* This is passed in as a list of #define lines, as generated by the C preprocessor with the
+     * appropriate switch to emit them. We append to it any #define's and #undef's encountered in the source
+     * file, as cpp with the -dD embeds them in the preprocessed output file.
+     * Once the file is parsed, then the #define's are converted to D symbols and appended to the array
+     * of Dsymbols returned by parseModule().
+     */
+    OutBuffer* defines;
```
Comment (maxhaton) [inline src/dmd/cparse.d:68]: Can this be an associative array?

File: src/dmd/cparse.d [hunk 6]
```diff
@@ -5090,4 +5112,63 @@ final class CParser(AST) : Parser!AST
...
+
+    /******************************************************************************/
+    /********************************* #define Parser *****************************/
+    //{
+
+    /**
+     * Go through the #define's in the defines buffer and see what we can convert
+     * to Dsymbols, which are then appended to symbols[]
+     */
+    void addDefines()
```
Comment (maxhaton) [inline src/dmd/cparse.d:5124]: This should really be a different file. It's going to get very complicated and should be using the C parser versus being part of the C parser.

File: src/dmd/cparse.d (hunk 5 above)
Comment (WalterBright) [reply src/dmd/cparse.d:68]: Since it is never searched, it wouldn't help.

File: src/dmd/cparse.d (hunk 6 above)
Comment (WalterBright) [reply src/dmd/cparse.d:5124]: I've toyed with the idea of making it a class derived from `CParser`. But it's a lot easier to make architectural decisions once we know what we're doing and how it's going to interact with `CParser`. So in the meantime I went for the most direct, straightforward approach.

Comment (WalterBright) [reply src/dmd/cparse.d:5124]: I still had to re-architect that section several times, even though it doesn't do much.

Comment (maxhaton) [reply src/dmd/cparse.d:5124]: It shouldn't be related by blood to CParser but rather be dependant on CParser, for manifest constants at least.

File: src/dmd/cparse.d (hunk 5 above)
Comment (maxhaton) [reply src/dmd/cparse.d:68]: I meant more in terms documentation. The name and comment isn't as descriptive as I'd like.

Comment (WalterBright) [reply src/dmd/cparse.d:68]: What is unclear about it? Please be more specific.

File: src/dmd/cparse.d (hunk 6 above)
Comment (WalterBright) [reply src/dmd/cparse.d:5124]: We won't really know what it needs from CParser until it is better developed. For example, some of the pattern matching can make use of "is the expansion a type", "is the expansion an expression", etc. It's much more than manifest constants.

Comment (WalterBright) [reply src/dmd/cparse.d:5124]: After all, the macro expansions are fragments of C syntax.

Comment (ibuclaw) [reply src/dmd/cparse.d:5124]: We already have mixin types and expressions in D, what won't work by just reusing them to represent macros?

Comment (ibuclaw) [reply src/dmd/cparse.d:5124]: > We already have mixin types and expressions in D, what won't work by just reusing them to represent macros? 

Yes, I'm well aware of the possibility to be tricksy with macros. :-)
```
// #define TYPE fun(); int
TYPE var;
mixin("fun(); int") var;
```

Comment (WalterBright) [reply src/dmd/cparse.d:5124]:
```
#define TYPE int long
```
How is that going to work with mixins? (I see similar macros cpp predefines.) I.e. how are you going to detect it is a type so you can generate the right mixin for it? I plan on using `CParser.isTypeName()` and those other `is` functions in CParser to detect what kind of thing the replacement text is.

**Comment by WalterBright** [comment]
I can't proceed with getting the macro definitions from `cl` and `sppn` until this is pulled.

File: src/dmd/cparse.d (hunk 6 above)
Comment (maxhaton) [reply src/dmd/cparse.d:5124]: This one should be trivially turned into a symbol directly but a mixin can be used in place of a type.

Also "generate the right mixin" sounds like a bad mixture of the two solutions. Mixin is a fallback, generation of symbols is supposed to be enough.

Comment (WalterBright) [reply src/dmd/cparse.d:5124]: Yeah, but you have to know it is a type in order to generate the right mixin. As for generating symbols, that's exactly what this PR does.

Comment (maxhaton) [reply src/dmd/cparse.d:5124]: We don't have to generate the right mixin as per se.

The mixins for types, decls, and expressions are done in different AST nodes, no?

**Comment by dkorpel** [comment]
The diff is full of 'Added lines were not covered by tests' warnings. I thought this might be win32 related again, but it turns out the test runner doesn't run .i files, so this isn't being tested at all.

**Review by dkorpel — APPROVED** [review]
I manually verified the test, but the test runner should be updated

**Pull request merged**